src/v2/tags.rs

```diff
@@ -5,65 +5,116 @@ use v2::*;
 
+/// A chunk of tags for an image.
+///
+/// This contains a non-strict subset of the whole list of tags
+/// for an image, depending on pagination option at request time.
 #[derive(Debug, Default, Deserialize, Serialize)]
-struct Tags {
+struct TagsChunk {
+    /// Image repository name.
     name: String,
+    /// Subset of tags.
     tags: Vec<String>,
 }
 
 impl Client {
     /// List existing tags for an image.
     pub fn get_tags(&self, name: &str, paginate: Option<u32>) -> StreamTags {
-        let url = {
-            let mut s = format!("{}/v2/{}/tags/list", self.base_url, name);
-            if let Some(n) = paginate {
-                s = s + &format!("?n={}", n);
+        let dclient = self.clone();
+        let base_url = format!("{}/v2/{}/tags/list", self.base_url, name);
+
+        let fres = futures::stream::unfold(Some(String::new()), move |link| {
+            // Stream ends when response has no `Link` header.
+            let last = match link {
+                None => return None,
+                Some(ref s) if s == "" => None,
+                s => s,
+            };
+
+            let full_url = match (paginate, last) {
+                (Some(p), None) => format!("{}?n={}", base_url, p),
+                (None, Some(l)) => format!("{}?next_page={}", base_url, l),
+                (Some(p), Some(l)) => format!("{}?n={}&next_page={}", base_url, p, l),
+                _ => base_url.to_string(),
             };
-            match hyper::Uri::from_str(s.as_str()) {
-                Ok(url) => url,
-                Err(e) => {
-                    return Box::new(futures::stream::once(Err(format!(
-                        "failed to parse url from string: {}",
-                        e
-                    ).into())));
-                }
-            }
-        };
-        let req = match self.new_request(hyper::Method::GET, url.clone()) {
-            Ok(r) => r,
-            Err(e) => {
-                let msg = format!("new_request failed: {}", e);
-                error!("{}", msg);
-                return Box::new(futures::stream::once(Err(msg.into())));
-            }
-        };
-        let freq = self.hclient.request(req);
-        let fres = freq
-            .from_err()
-            .and_then(|r| {
-                let status = r.status();
-                if status != hyper::StatusCode::OK {
-                    return Err(format!("get_tags: wrong HTTP status '{}", status).into());
-                };
-                {
-                    let ct_hdr = r.headers().get(header::CONTENT_TYPE);
+            let client = dclient.clone();
+            let url = hyper::Uri::from_str(&full_url);
+
+            let freq = futures::future::result(url)
+                .from_err()
+                .and_then(move |url| {
+                    trace!("GET {:?}", &url);
+                    let req = client.new_request(hyper::Method::GET, url);
+                    futures::future::result(req)
```

Comment:
All these different future combinators look messy here. I've failed a quick try to reduce the nesting level down to one, but I'd really like to see that for readability.

Reply:
This code is equivalent to https://github.com/camallo/dkregistry-rs/blob/23adf75204f96d8609ec1534bdefea63f678d9d8/src/v2/auth.rs#L27-L46

It seems quite concise/idiomatic, but I understand it's a subjective point.